### PR TITLE
fix: Use xtime duration to parse batch job

### PR DESCRIPTION
**Pull request by jiuker**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
fix: Use xtime to parse batch job
let batchJob yaml support:
`olderThan: 7d10h`
or 
`olderThan: 1w`

## How to test this PR?

```yaml
replicate:
  apiVersion: v1
  # source of the objects to be replicated
  source:
    type: minio # valid values are "s3" or "minio"
    bucket: mytest
    prefix: object-prefix1 # 'PREFIX' is optional
    # If your source is the 'local' alias specified to 'mc batch start', then the 'endpoint' and 'credentials' fields are optional and can be omitted
    # Either the 'source' or 'remote' *must* be the "local" deployment
#    endpoint: "http://127.0.0.1:9000"
#    # path: "on|off|auto" # "on" enables path-style bucket lookup. "off" enables virtual host (DNS)-style bucket lookup. Defaults to "auto"
#    credentials:
#      accessKey: minioadmin # Required
#      secretKey: minioadmin # Required
#    # sessionToken: SESSION-TOKEN # Optional only available when rotating credentials are used
    snowball: # automatically activated if the source is local
      disable: true # optionally turn-off snowball archive transfer
#      batch: 100 # upto this many objects per archive
#      inmemory: true # indicates if the archive must be staged locally or in-memory
#      compress: false # S2/Snappy compressed archive
#      smallerThan: 5MiB # create archive for all objects smaller than 5MiB
#      skipErrs: false # skips any source side read() errors

  # target where the objects must be replicated
  target:
    type: minio # valid values are "s3" or "minio"
    bucket: mytest
    prefix: stage # 'PREFIX' is optional
    # If your source is the 'local' alias specified to 'mc batch start', then the 'endpoint' and 'credentials' fields are optional and can be omitted

    # Either the 'source' or 'remote' *must* be the "local" deployment
    endpoint: "http://127.0.0.1:9001"
    # path: "on|off|auto" # "on" enables path-style bucket lookup. "off" enables virtual host (DNS)-style bucket lookup. Defaults to "auto"
    credentials:
      accessKey: minioadmin
      secretKey: minioadmin
    # sessionToken: SESSION-TOKEN # Optional only available when rotating credentials are used

  # NOTE: All flags are optional
  # - filtering criteria only applies for all source objects match the criteria
  # - configurable notification endpoints
  # - configurable retries for the job (each retry skips successfully previously replaced objects)
  flags:
    filter:
      newerThan: "7d10h31s" # match objects newer than this value (e.g. 7d10h31s)
      olderThan: "1w" # match objects older than this value (e.g. 7d10h31s)
#      createdAfter: "date" # match objects created after "date"
#      createdBefore: "date" # match objects created before "date"

      ## NOTE: tags are not supported when "source" is remote.
      tags:
         - key: "name"
           value: "pick*" # match objects with tag 'name', with all values starting with 'pick'

      metadata:
         - key: "content-type"
           value: "image/*" # match objects with 'content-type', with all values starting with 'image/'

#    notify:
#      endpoint: "https://notify.endpoint" # notification endpoint to receive job status events
#      token: "Bearer xxxxx" # optional authentication token for the notification endpoint
#
#    retry:
#      attempts: 10 # number of retries for the job before giving up
#      delay: "500ms" # least amount of delay between each retry


```

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/batch-expire.go [hunk 1]
```diff
@@ -131,13 +132,33 @@ var _ yaml.Unmarshaler = &BatchJobExpireFilter{}
 // UnmarshalYAML - BatchJobExpireFilter extends unmarshal to extract line, col
 // information
 func (ef *BatchJobExpireFilter) UnmarshalYAML(value *yaml.Node) error {
-	type expFilter BatchJobExpireFilter
-	var tmp expFilter
+	tmp := struct {
```
Comment (klauspost) [inline cmd/batch-expire.go:135]: If you have modified `(*BatchReplicateFilter).UnmarshalYAML` why do another translation here?

Wouldn't the original code work?

**Comment by klauspost** [comment]
Instead of all this dancing back and forth I added msgp serialization to xtime.Duration: https://github.com/minio/pkg/pull/122

This should mean you can just replace the struct types.

**Comment by klauspost** [comment]
(at least I am guessing that is why you have all this conversion)

**Comment by jiuker** [comment]
@klauspost this pr is for yaml unmarshal. Yesh. It's ugly. Inline struct will not work for yaml. That's why another pr is draft so far.

**Comment by klauspost** [comment]
@jiuker Please explain why you need to unmarshal into a temporary struct and copy fields manually?

> That's why another pr is draft so far.

I don't understand.

Can't this mostly just be:

```
-	OlderThan     time.Duration       `yaml:"olderThan,omitempty" json:"olderThan"`
+	OlderThan     xtime.Duration      `yaml:"olderThan,omitempty" json:"olderThan"`
```

**Comment by jiuker** [comment]
> @jiuker Please explain why you need to unmarshal into a temporary struct and copy fields manually?
> 
> > That's why another pr is draft so far.
> 
> I don't understand.
> 
> Can't this mostly just be:
> 
> ```
> -	OlderThan     time.Duration       `yaml:"olderThan,omitempty" json:"olderThan"`
> +	OlderThan     xtime.Duration      `yaml:"olderThan,omitempty" json:"olderThan"`
> ```

It can. But so many others will do type translate. And need the pr that you commited

**Comment by klauspost** [comment]
Added this... dunno if that helps:
```
// D will return as a time.Duration.
func (d Duration) D() time.Duration {
	return time.Duration(d)
}
```

**Review by krisis — APPROVED** [review]

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Upgrade the pkg to use the new marshal technique

**Comment by krisis** [comment]
I didn't realize that pkg/xtime was making the xtime.Duration to time.Duration values unnecessary. Barring that, the PR looks good to me.

**Review by harshavardhana — APPROVED** [review]

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Why is temporary struct needed?

File: cmd/batch-handlers.go [hunk 2]
```diff
@@ -287,12 +287,12 @@ func (r *BatchJobReplicateV1) StartFromSource(ctx context.Context, api ObjectLay
 	isStorageClassOnly := len(r.Flags.Filter.Metadata) == 1 && strings.EqualFold(r.Flags.Filter.Metadata[0].Key, xhttp.AmzStorageClass)
 
 	skip := func(oi ObjectInfo) (ok bool) {
-		if r.Flags.Filter.OlderThan > 0 && time.Since(oi.ModTime) < r.Flags.Filter.OlderThan {
+		if r.Flags.Filter.OlderThan > 0 && time.Since(oi.ModTime) < time.Duration(r.Flags.Filter.OlderThan) {
```
Comment (harshavardhana) [inline cmd/batch-handlers.go]:
```suggestion
		if r.Flags.Filter.OlderThan > 0 && time.Since(oi.ModTime) < r.Flags.Filter.OlderThan.D() {
```

**Comment by jiuker** [comment]
> Added this... dunno if that helps:
> 
> ```
> // D will return as a time.Duration.
> func (d Duration) D() time.Duration {
> 	return time.Duration(d)
> }
> ```

Yesh. It help

**Review by ramondeklein — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**